OWLS-87849 - scaleCluster.sh script to patch the 'replicas' attribute of the WLS cluster

ankedia opened the pull request:
OWLS-87849 - The `scaleCluster.sh` script is similar to the cluster stop/start scripts and it patches the replicas attribute for the cluster with the specified value. It performs validations to ensure that clusterName exists in domain topology and that the specified value of replica count is in the allowed range. If the cluster doesn't exist under `.spec.clusters` in the domain resource, it adds the cluster. 
The PR also changes the logic for the cluster stop/start scripts to add the cluster under `.spec.clusters` in the domain resource if it doesn't exist.

Integration test run URL - https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/4171/

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/helper.sh:
```diff
@@ -478,6 +547,24 @@ function getMinReplicas {
+  eval $__result=0
+  maxReplicaCmd="(.status.clusters[] | select (.clusterName == \"${clusterName}\")) \
+    | .maximumReplicas"
+  maxReplicas=$(echo ${domainJson} | jq "${maxReplicaCmd}")
```
please declare maxReplicaCmd & maxReplicas as local

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/helper.sh:
```diff
@@ -478,6 +547,24 @@ function getMinReplicas {
+  local clusterName=$2
+  local __result=$3
+
+  eval $__result=0
```
What is the purpose of this eval? The next eval will replace __result...

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/helper.sh:
```diff
@@ -478,6 +547,24 @@ function getMinReplicas {
+  maxReplicaCmd="(.status.clusters[] | select (.clusterName == \"${clusterName}\")) \
+    | .maximumReplicas"
+  maxReplicas=$(echo ${domainJson} | jq "${maxReplicaCmd}")
+  eval $__result=${maxReplicas}
```
Should this be ${maxReplicas:-0} ?

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/helper.sh:
```diff
@@ -478,6 +547,24 @@ function getMinReplicas {
+  maxReplicaCmd="(.status.clusters[] | select (.clusterName == \"${clusterName}\")) \
+    | .maximumReplicas"
+  maxReplicas=$(echo ${domainJson} | jq "${maxReplicaCmd}")
+  eval $__result=${maxReplicas}
```
there is a problem here if $__result happens to be maxReplicas -- since 'maxReplicas' should be a local var the two will complete -- choose something more unique for the local var...

tbarnes-us replied:
overall, it'd probably be a nice 'code cleanup' to update every function's local vars to use an __ pattern - this would eliminate the potential for them to collide with globals...

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/helper.sh:
```diff
@@ -460,6 +503,32 @@ function isReplicaCountEqualToMinReplicas {
+  if [ ${replicas} -lt ${minReplicas} ] || [ ${replicas} -gt ${maxReplicas} ]; then
+    eval $__result=false
+    eval $__range="'${range}'"
+  fi
```
(1) It looks like range is supposed to be a local var.  Make it local and give it a unique name that doesn't overlap with any global.
(2) It looks like there's a missing 'eval' for $__range if we fall into the if-then stanza

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/helper.sh:
```diff
@@ -249,13 +249,56 @@ function createPatchJsonToUpdateClusterPolicy {
+      .[\$idx][\"serverStartPolicy\"] = \"${policy}\" \
+      else .+  [{clusterName: \"${clusterName}\" , serverStartPolicy: \"${policy}\"}] end"
+    startPolicy=$(echo ${clusters} | jq "${mapCmd}")
+  fi
```
Why is this change here? It looks like it relates to 'start policy' not 'replicas', but this pull is all about replicas.  Maybe a bad merge?

tbarnes-us replied:
Also, there's a lot of vars here that should be declared local...

ankedia replied:
It's not a bad merge. If the cluster name is not under `.spec.clusters` for the `stopCluster.sh` or `startCluster.sh` script then it'll not work correctly. Previously I was assuming that cluster will always be present under `.spec.clusters`. Since we removed that assumption for the `scaleCluster.sh` script, I'm making a similar change for `stopCluster.sh` and `startCluster.sh` scripts.

tbarnes-us replied:
OK. Should this be back-ported?

ankedia replied:
SOA team has tested the cluster start/stop scripts and didn't run into any issues with the existing version in 3.1.4. My understanding is that it's an extremely rare use-case to not have any cluster defined under `.spec.clusters` in domain resource. There's a workaround to add the cluster with IF_NEEDED serverStartPolicy. IMO it's not a very high priority to backport since SOA team is also planning to move to 3.2.0 release in May/June.

tbarnes-us replied:
> My understanding is that it's an extremely rare use-case to not have any cluster defined under .spec.clusters in domain resource. 

I think it depends on whether this is during in development or production. I suspect it's the more common case in development.

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/helper.sh:
```diff
@@ -249,13 +249,56 @@ function createPatchJsonToUpdateClusterPolicy {
+      else .+  [{clusterName: \"${clusterName}\" , replicas: ${replicas}}] end"
+    replicasPatch=$(echo ${clusters} | jq "${mapCmd}")
+  fi
+  patchJson="{\"spec\": {\"clusters\": "${replicasPatch}"}}"
```
addClustersReplicasCmd, mapCmd, replicasPath, and patchJson should all be declared as local.  They also should be sufficiently uniquely named so that it's unlikely their names can collide with the value that might be in "$__result".

overall, it might be good overall pattern to have all 'local' variables start with __, and then avoid using __ for global vars...

ankedia replied:
I'm using the __ pattern for the return parameters. For now, I'll make sure that local parameters are not used in the global scope.

tbarnes-us replied:
Thinking about this some more, I assume they occlude if they're declared local - which would make them safe...

tbarnes-us replied:
... as long as they're declared local...

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/scaleCluster.sh:
```diff
@@ -0,0 +1,123 @@
+ 
+    $(basename $0) -c mycluster -r replicas [-n mynamespace] [-d mydomainuid] [-m kubecli]
+  
+    -c <cluster-name>   : Cluster name parameter is required.
```
Can we consider defaulting this to 'cluster-1'? That's what we use in the samples...

ankedia replied:
We don't default cluster-name to `cluster-1` for `stopCluster.sh` and `startCluster.sh` scripts. If there are multiple clusters in the domain and the user intended to change replicas for `cluster-2` but forgot to specify the cluster-name parameter then it might cause unintended behavior.

tbarnes-us replied:
Ok.

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/helper.sh:
```diff
@@ -478,6 +547,24 @@ function getMinReplicas {
   eval $__result=${minReplicas}
```
Like the other commands, minReplicaCmd & minReplicas should be declared local. Furthermore, they should be unique enough so that there's potential collision with $__result...

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/scaleCluster.sh:
```diff
@@ -0,0 +1,123 @@
+# !/bin/sh
+# Copyright (c) 2020, 2021, Oracle and/or its affiliates.
```
Technically, you can remove the '2020,' since this is new code for 2021.

tbarnes-us commented on kubernetes/samples/scripts/domain-lifecycle/scaleCluster.sh:
```diff
@@ -0,0 +1,123 @@
+
+  cat << EOF
+
+  This script scales a WebLogic cluster in a domain by patching
```
by patching --> by patching the

ankedia commented:
@tbarnes-us I have addressed the review comments, please let me know if I missed anything. I also ran the integration tests with the changes and test run passed - https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/4171 . Thanks.

tbarnes-us reviewed (approved):
LGTM. See also comment/question about back-porting due to the included fix.
